pyteal/ast/router.py

```diff
@@ -259,13 +263,203 @@ class CondNode:
+        """This method filters out invalid handlers that might be normal subroutine, Expr, or unroutable ABIReturnSubroutine.
+
+        It accepts only routable ABIReturnSubroutine, and shrink the type to ABIReturnSubroutine from argument's union type.
+        """
+        if not isinstance(subroutine, ABIReturnSubroutine):
+            raise TealInputError(
+                f"method call should be only registering ABIReturnSubroutine, got {type(subroutine)}."
+            )
+        if not subroutine.is_abi_routable():
+            raise TealInputError(
+                f"method call ABIReturnSubroutine is not routable: "
+                f"got {subroutine.subroutine.argument_count()} args "
+                f"with {len(subroutine.subroutine.abi_args)} ABI args."
+            )
+        return subroutine
+
+    @staticmethod
+    def __subroutine_argument_instance_generate(
+        subroutine: ABIReturnSubroutine,
+    ) -> tuple[list[abi.BaseType], list[abi.BaseType], list[abi.Transaction]]:
+        # All subroutine args types
+        type_specs = cast(list[abi.TypeSpec], subroutine.subroutine.expected_arg_types)
+
+        # All subroutine arg values, initialize here and use below instead of
+        # creating new instances on the fly, so we don't have to think about splicing
+        # back in the transaction types
+        arg_vals = [typespec.new_instance() for typespec in type_specs]
+
+        # Only args that appear in app args
+        app_arg_vals: list[abi.BaseType] = [
+            ats for ats in arg_vals if not isinstance(ats, abi.Transaction)
+        ]
+
+        # only transaction args (these are omitted from app args)
+        txn_arg_vals: list[abi.Transaction] = [
+            ats for ats in arg_vals if isinstance(ats, abi.Transaction)
+        ]
+
+        for aav in app_arg_vals:
+            # If we're here we know the top level isn't a Transaction but a transaction may
+            # be included in some collection type like a Tuple or Array, raise error
+            # as these are not supported
+            if abi.contains_type_spec(aav.type_spec(), abi.TransactionTypeSpecs):
+                raise TealInputError(
+                    "A Transaction type may not be included in Tuples or Arrays"
+                )
+
+        return arg_vals, app_arg_vals, txn_arg_vals
+
+    @staticmethod
+    def __decode_constructions_and_args(
+        arg_vals: list[abi.BaseType],
+        app_arg_vals: list[abi.BaseType],
+        txn_arg_vals: list[abi.Transaction],
+        subroutine: ABIReturnSubroutine,
+        use_frame_pt: bool = False,
+    ) -> tuple[list[Expr], list[abi.BaseType], Optional[Proto]]:
+        """
+        Assumption: arg_vals = app_args_vals union with txn_arg_vals
+        """
+
+        # if subroutine has ABI output, then local variables start from 1
+        # otherwise local variables start from 0
+        index_start_from = int(subroutine.output_kwarg_info is not None)
```

Comment:
because it is returning an index
maybe `index_start_from = 0 if subroutine.output_kwarg_info is None else 1`

Reply:
[7ade42b](https://github.com/algorand/pyteal/pull/600/commits/7ade42b68df7399ce389d9891265192b2d0e5502)